worker/backend/backend.go

```diff
@@ -74,225 +109,176 @@ func (b *Backend) Stop() {
 	if err != nil {
-		err = ClientError{InnerError: fmt.Errorf("failed to convert garden spec to oci spec: %w", err)}
-		return
+		return nil, fmt.Errorf("garden spec to oci spec: %w", err)
+	}
+
+	netMounts, err := b.network.SetupMounts(gdnSpec.Handle)
+	if err != nil {
+		return nil, fmt.Errorf("network setup mounts: %w", err)
 	}
 
-	cont, err := b.client.NewContainer(ctxWithTimeout,
-		gdnSpec.Handle, gdnSpec.Properties, oci,
-	)
+	oci.Mounts = append(oci.Mounts, netMounts...)
 
+	cont, err := b.client.NewContainer(ctx, gdnSpec.Handle, gdnSpec.Properties, oci)
 	if err != nil {
-		err = ClientError{InnerError: fmt.Errorf("failed to create a container in containerd: %w", err)}
-		return
+		return nil, fmt.Errorf("new container: %w", err)
 	}
 
-	_, err = cont.NewTask(ctxWithTimeout, cio.NullIO)
+	task, err := cont.NewTask(ctx, cio.NullIO)
 	if err != nil {
-		err = ClientError{InnerError: fmt.Errorf("failed to create a task in container: %w", err)}
-		return
+		return nil, fmt.Errorf("new task: %w", err)
 	}
 
-	container = &Container{
-		handle: cont.ID(),
+	err = b.network.Add(ctx, task)
+	if err != nil {
+		return nil, fmt.Errorf("network add: %w", err)
 	}
-	return
+
+	err = task.Start(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("task start: %w", err)
+	}
+
+	return NewContainer(
+		cont,
+		b.containerStopper,
+		b.rootfsManager,
+	), nil
 }
 
-// Destroy destroys a container.
-//
-// When a container is destroyed, its resource allocations are released,
-// its filesystem is removed, and all references to its handle are removed.
-//
-// All resources that have been acquired during the lifetime of the container are released.
-// Examples of these resources are its subnet, its UID, and ports that were redirected to the container.
-//
-// TODO: list the resources that can be acquired during the lifetime of a container.
-//
-// Errors:
-// * Container handle does not exist
-// * Container task was not successfully spun down prior to container deletion.
-//   Reasons include task not found, termination signal failed, or task deletion failed.
-// * Destroy request to client failed
+// Destroy gracefully destroys a container.
 //
 func (b *Backend) Destroy(handle string) error {
 	if handle == "" {
-		return InputValidationError{Message: "handle is empty"}
+		return ErrInvalidInput("empty handle")
 	}
 
-	ctx := namespaces.WithNamespace(context.Background(), b.namespace)
-	ctxWithTimeout, _ := context.WithTimeout(ctx, b.clientTimeout)
+	ctx := context.Background()
+
+	container, err := b.client.GetContainer(ctx, handle)
+	if err != nil {
+		return fmt.Errorf("get container: %w", err)
+	}
 
-	container, err := b.client.GetContainer(ctxWithTimeout, handle)
+	err = b.containerStopper.GracefullyStop(ctx, container)
 	if err != nil {
-		return ClientError{InnerError: err}
+		return fmt.Errorf("stopping container: %w", err)
 	}
 
-	task, err := container.Task(ctxWithTimeout, nil)
+	task, err := container.Task(ctx, cio.Load)
```

Comment:
This will work even after the init process is killed, right? (`GracefullyStop` eventually kills it)

Reply:
oooooh good catch! the "trying to load the task" would "work" - it'd always return the "not found" that we catch right after the `if err`, which means that in practice we'd always skip the network tear down :sweat_smile:

Reply:
refactored the whole thing a bit to let us perform the deletion not during the killing phase (it was the "containerstopper" responsability), but instead, do after the network teardown is performed.